aries_vcx/src/handlers/connection/inviter/state_machine.rs

```diff
@@ -240,38 +243,36 @@ impl SmConnectionInviter {
         new_pw_vk: &str,
         send_message: fn(&str, &DidDoc, &A2AMessage) -> VcxResult<()>,
     ) -> VcxResult<()> {
-        send_message(&new_pw_vk, &state.did_doc, &state.signed_response.to_a2a_message())?;
-        Ok(())
+        send_message(&new_pw_vk, &state.did_doc, &state.signed_response.to_a2a_message())
     }
 
     pub fn handle_connect(self, routing_keys: Vec<String>, service_endpoint: String) -> VcxResult<Self> {
-        let Self { source_id, pairwise_info, state, .. } = self;
-        let state = match state {
+        let state = match self.state {
             InviterFullState::Initial(state) => {
                 let invite: PairwiseInvitation = PairwiseInvitation::create()
-                    .set_label(source_id.to_string())
-                    .set_recipient_keys(vec!(pairwise_info.pw_vk.clone()))
+                    .set_id(&self.thread_id)
+                    .set_label(&self.source_id)
+                    .set_recipient_keys(vec!(self.pairwise_info.pw_vk.clone()))
                     .set_routing_keys(routing_keys)
                     .set_service_endpoint(service_endpoint);
 
                 InviterFullState::Invited(
                     (state, Invitation::Pairwise(invite)).into()
                 )
             }
-            _ => {
-                state.clone()
-            }
+            _ => self.state.clone()
         };
-        Ok(Self { source_id, pairwise_info, state, ..self })
+        Ok(Self { state, ..self })
     }
 
     pub fn handle_connection_request(self,
                                      request: Request,
                                      new_pairwise_info: &PairwiseInfo,
                                      new_routing_keys: Vec<String>,
                                      new_service_endpoint: String) -> VcxResult<Self> {
-        let Self { pairwise_info: bootstrap_pairwise_info, state, send_message, .. } = self.clone();
-        let state = match state {
+        let bootstrap_pairwise_info = self.pairwise_info.clone();
+        let thread_id = request.get_thread_id().ok_or(VcxError::from_msg(VcxErrorKind::InvalidJson, "Missing ~thread decorator field in request"))?;
```

Comment:
Also I think we miss matching the thread id from the request against our known `self.thread_id`?

Reply:
Oh I see we do that in `build_response` - I left a comment there and consquently if we always do thread verification inside `build_response`, we don't have to check if request contains `thread_id` at this place

Reply:
I agree it seems confusing, maybe we should consolidate `handle_connection_request` and `handle_send_response` (i.e. `Requested` and `Responded` states) into one step.

Reply:
Yeah I think that's a good idea, that way connection response can be constructed arbitrarily on upper layers without forcing a particular way on SM layer

Reply:
But back to this particular line of code - especially with the changes you did in `build_response`, I think this line can be completely removed - if `thread_id` is missing in request, it will be detected in `build_response` and problem report will be sent (so in that way it's even more desired behaviour, doesn't happen with line `274`)